### ISSUE TITLE
jwt_verify with ES256 / ES512 does not work

### ISSUE THREAD

**Issue by EtienneBruines** [haproxy/haproxy#2001]
### Detailed Description of the Problem

`jwt_verify` keeps returning `0`, despite it being a valid signature.

```
var(txn.bearer),jwt_verify(txn.alg,"/tmp/pk.pem")
```

This is the case for ES256, ES384 and ES512 algorithms. RS256 seems to work. 

### Expected Behavior

`jwt_verify` should return `1` instead.

Or at least a `-1` or `-2` if it cannot handle this algorithm. 

### Steps to Reproduce the Behavior

1. Go to https://jwt.io/ and select `ES256`
2. Copy the public key that was generated into a file
3. Run haproxy with the correct path to that public key file (in the `jwt_verify`)
4. Use `curl 127.0.0.1:8080 -v -H "Authorization: Bearer paste-signature-here"`
5. Observe the result

### What is your configuration?

```haproxy
global
    master-worker
    log          127.0.0.1 local2 debug
    stats socket 127.0.0.1:14567
    #stats socket /tmp/haproxy.sock mode 666 level admin
    log          stdout local0 debug
    maxconn      2000

defaults
    mode                    http
    log                     global
    option                  httpslog
    option                  dontlognull
    option                  http-keep-alive
    option                  forwardfor
    option                  redispatch

    timeout http-request    30s
    timeout queue           1m
    timeout connect         3s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 2000
    default_backend dummy_backend

backend dummy_backend
    server server_static1 127.0.0.1:8000

  frontend default_frontend
    bind *:8080

    http-request set-var(txn.bearer) http_auth_bearer if { http_auth_bearer -m found }

    # Deny all requests to /api/ that do not have an bearer value
    http-request deny deny_status 401 content-type 'application/json' string '{"error": "Missing Authorization HTTP header"}' if !{ var(txn.bearer) -m found }

    # get header part of the JWT
    http-request set-var(txn.alg) var(txn.bearer),jwt_header_query('$.alg')

    # To output the actual return value of jwt_verify:
    http-after-response set-header X-Debug %[var(txn.bearer),jwt_verify(txn.alg,"/tmp/pk.pem")]

    # Validate JWT
    http-request deny content-type 'application/json' string '{"error": "Unsupported JWT signing algorithm"}' if !{ var(txn.alg) -m str ES256 }
    http-request deny content-type 'application/json' string '{"error": "Invalid JWT signature"}' if !{ var(txn.bearer),jwt_verify(txn.alg,"/tmp/pk.pem") -m int 1 }
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.7.1-3e4af0e 2022/12/19 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.1.html
Running on: Linux 5.15.0-58-generic #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -PTHREAD_EMULATION +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL -OPENSSL_WOLFSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING +SHM_OPEN

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Additional Information

Using the docker image `haproxy:2.7.1` but also occurs on any other version (e.g. `2.7.1-alpine` or `2.6.4` or `2.5.1`), as well as non-docker builds. 

### Bug-Hunting

It seems like [this line](https://github.com/haproxy/haproxy/blob/master/src/jwt.c#L267) returns `-1`:

```c
EVP_DigestVerifyFinal(evp_md_ctx, (const unsigned char*)decoded_signature->area, decoded_signature->data)
```

**Comment by rlebreton** [comment]
Thanks for your report, I'll look into it. The ecdsa tokens are tested by one of our regression tests so it's surprising that it would fail for a given certificate pair but work for others. There must be something I missed when implementing it.

**Comment by EtienneBruines** [comment]
There might be some issues where the JWT signature is not in the `ANS.1` format that OpenSSL likes to work with, but I haven't been able to confirm that theory. 

(And since the JWT spec as well as the `jwt_verify` both expect the not-ASN1 format, it's difficult to convert and test it out.)

The regression test `jws_verify.vtc` works, but as soon as I regenerate the keys using the commands provided, the JWT validates on https://jwt.io/ but the haproxy test then does fail:

```
########################## Gathering results ##########################
###### Test case: reg-tests/jwt/jws_verify.vtc ######
## test results in: "/tmp/haregtests-2023-01-16_14-29-59.OXn2BX/vtc.90476.02dda4ea"
---- c9    EXPECT resp.http.x-jwt-verify-ES256 (0) == "1" failed
*    diag  0.0 /usr/bin/socat
make: *** [Makefile:1153: reg-tests] Error 1
```

# Update

The signature from Hashicorp Vault was exactly 64 bytes, which made me think of the 2x32 integer values that the signature consists of. After [manually converting it to ASN1 DER](https://stackoverflow.com/a/59905274/3243814) both openssl and haproxy are happy with the result.

The signature generated by `openssl dgst -sha256 -sign es256-private.pem data.txt` as well as the signature in the `jws_verify.vtc` file are also definitely in ASN1 format, which is not according to [spec](https://www.rfc-editor.org/rfc/rfc7518#section-3.4). That doesn't explain though, why the newly generated signature isn't accepted by the tests (even though the hand-crafted one from Hashicorp Vault did manage to work). 

## Update 2

Testing is HARD. Especially if there's a `\r` hiding somewhere that doesn't visually show. Whoops... 

Generating new keys "works" - the regression test passes.

The main problem is, that the signature should not be in ASN1 format, but instead have the two 32-byte integer values appended to each other - according to spec. The tests pass when provided a signature in the ASN1 format, but not using the "raw" base64url format. 

# Solution

Checking for ECDSA-signatures if the length is exactly 64 bytes - if so, [convert it to ASN1](https://stackoverflow.com/a/59905274/3243814) and feed that to OpenSSL. 

I won't be able to program that in C without memory leaking, though. Hope this helps. 

https://manpages.ubuntu.com/manpages/bionic/man3/ECDSA_SIG_new.3ssl.html
```
       i2d_ECDSA_SIG() creates the DER encoding of the ECDSA signature sig and writes the encoded
       signature to *pp (note: if pp is NULL i2d_ECDSA_SIG() returns the expected length in bytes
       of the DER encoded signature).  i2d_ECDSA_SIG() returns the length of the DER encoded
       signature (or 0 on error).
```

**Comment by rlebreton** [comment]
Yes you are right, the implementation is indeed wrong. I did not test the ESXXX implementation correctly and assumed that OpenSSL's output format complied with the RFC and only needed to be converted to base64 as-is. The verify part of the code was also built with this assumption so it needs to be reworked. 
I'll try to come up with a proper implementation asap. Thanks for your detailed analysis, I'll come up with a patch that will hopefully end up on the branch you use soon.

**Comment by wlallemand** [comment]
The patch was merged for 2.7.2, unfortunately https://github.com/haproxy/haproxy/issues/2005 was merged only after the release.

**Comment by wtarreau** [comment]
At least it's in the branch, in case it helps to just pick that patch on top of the official version.
